File: Plus One Linked List.py
Problem link on lintcode:https://www.lintcode.com/problem/904/
Problem link on leetcode:https://leetcode.com/problems/plus-one-linked-list/



"""
Definition of ListNode
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""

class Solution:
    """
    @param head: the first Node
    @return: the answer after plus one
    """
    def plusOne(self, head):
        d=dummynode=ListNode(0)
        dummynode.next=head
        notnine=ListNode(0)
        temp=d
        while d!=None:
            if d.val!=9:
                notnine=d
            d=d.next
        notnine.val+=1
        notnine=notnine.next
        while notnine!=None:
            notnine.val=0
            notnine=notnine.next
        if dummynode.val==0:
            return temp.next
        else:
            return temp

File: Plus One Linked List.java
// Problem link on lintcode:https://www.lintcode.com/problem/904/
// Problem link on leetcode:https://leetcode.com/problems/plus-one-linked-list/

/**
 * Definition for ListNode
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution {
    private ListNode reverse(ListNode head){
        ListNode prev = null, curr = head;
        while(curr != null){
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
    public ListNode plusOne(ListNode head) {
        head = reverse(head);
        ListNode ans = new ListNode(0);
        ListNode curr = ans;
        int carry = 1;
        while(head!=null){
            int s = head.val + carry;
            if( s > 9){
                carry = s/10;
                s = s%10;
            }
            else{
                carry = 0;
            }
            curr.next = new ListNode(s);
            curr = curr.next;
            head = head.next;
        }
        if(carry > 0)
            curr.next = new ListNode(carry);
        return reverse(ans.next);
    }
}
